### ISSUE TITLE
[Question] User graph embedding

### ISSUE THREAD

**Issue by mbaddar** [facebookresearch/PyTorch-BigGraph#46]
I am training a user identity graph containing user events and interactions. Nodes have two types: Users and properties. Properties represent user cookies, IP address, pageviews etc. I am not sure if that would require featurized entities as so far I am not able to converge on a small dataset of around 300K edges.

I guess there is something fundamentally missing on my setup and I would like to have any troubleshooting guidance on how to converge. 

 Most of the graph I'm training on consists of isolated nodes with some properties. So they are not connected to the rest of the graph. 

Here is the result of 60 epochs: 
loss 0.0132791. violators_lhs:  0.00924837 , violators_rhs:  0.00964025 
pos_rank:  482.742 , mrr:  0.0115865 , r1:  0.00390625 , r10:  0.0117188 , r50:  0.0664062 , auc:  0.765625

Thanks a lot

**Comment by lerks** [comment]
I'm not sure exactly what you are asking for.

If you're wondering whether graph embeddings as produced by PBG are the right approach to solve your problem, well, this is really a design decision of your system that you have to make. We can give some pointers but ultimately this type of consulting isn't what these GitHub issues are for. Also, for anyone to help you on this, you should tell exactly what you're trying to achieve (what do you want the embeddings to capture) and what your graph represents (e.g., why are there disconnected nodes and what do you hope to learn from them?). You may also want to consider other approaches, such as graph neural networks, if you need to process features or labels of your nodes.

If you're having technical issues with PBG itself (e.g., training not converging), then this is the right place to get support for them, but more information is needed. At least your full config, ideally also your input data (if small enough and not private), and some more info on your setup. General advice is that, when writing a config, you should basically stick to models from the literature that are known to work (some are referenced in the documentation). Only if these models fail (and you really know what you're doing) you should to come up with a new model and hyperparameters. In this latter case, you'll probably have to do some grid search or other optimization on the learning rate and such to find the best setup.

**Comment by adamlerer** [comment]
@mbaddar can you provide some more information about the statistics of your graph?
- how many nodes are in the graph?
- how many edges does each node have? Do they all have the same number of edges or is there a big range? If nodes have very few edges then graph embeddings can't be used
- what do the edges represent? Would you expect that looking at some of the edges from a node you could predict other edges? MRR will stay low if edges don't provide any information that helps you predict other edges.

**Comment by mbaddar** [comment]
Thank you for your answers @adamlerer and @lerks 
The graph I'm trying to embed represent user identities. Users who we believe are similar to each other are connected by an 'alias' edge and they all connect to properties via 'has' edge

 I have tried different number of nodes. The last run was on around 288K edges and 314K nodes. 
Some user nodes only have a few edges (as they are not similar to any other user and do not have that many properties). And some others have more. 
I guess by looking at edges connecting similar users we could infer a similarity heuristic and apply it to other user pairs. This is what I was hoping to achieve anyways. 

Attached a graph sample and my config. 

[PBG.zip](https://github.com/facebookresearch/PyTorch-BigGraph/files/3178801/PBG.zip)

**Comment by adamlerer** [comment]
@mbaddar graph embeddings typically only work if you have >>1 edges/vertex. In your graph most nodes have 1 or even 0 edges, so the embeddings can't possibly encode anything useful. You should either find a way to add more information (more edges) to the graph (~50 edges/node is a good rule of thumb) or consider a different method.

**Comment by lerks** [comment]
There has been no activity here for a week (and I have nothing to add to what @adamlerer said), so I'm closing this. Please reopen if there's more you want to know/discuss.
